packages/web-client/app/components/card-pay/issue-prepaid-card-workflow/preview/index.ts

```diff
@@ -29,26 +29,49 @@ export default class CardPayDepositWorkflowPreviewComponent extends Component<Ca
   @reads('args.workflowSession.state.spendFaceValue')
   declare faceValue: number;
 
-  @tracked errorMessage = '';
+  @tracked error = '';
 
   @task *issueTask(): TaskGenerator<void> {
+    this.error = '';
+
     let { workflowSession } = this.args;
-    yield this.hubAuthentication.ensureAuthenticated();
-    // yield statements require manual typing
-    // https://github.com/machty/ember-concurrency/pull/357#discussion_r434850096
-    let customization: Resolved<PrepaidCardCustomization> = yield taskFor(
-      this.cardCustomization.createCustomizationTask
-    ).perform({
-      issuerName: workflowSession.state.issuerName,
-      colorSchemeId: workflowSession.state.colorScheme.id,
-      patternId: workflowSession.state.pattern.id,
-    });
-    yield taskFor(this.layer2Network.issuePrepaidCard)
-      .perform(this.faceValue, customization.did)
-      .then((address: string) => {
-        this.args.workflowSession.update('prepaidCardAddress', address);
-        this.args.onComplete();
+    try {
+      yield this.hubAuthentication.ensureAuthenticated();
+
+      // yield statements require manual typing
+      // https://github.com/machty/ember-concurrency/pull/357#discussion_r434850096
+      let customization: Resolved<PrepaidCardCustomization> = yield taskFor(
+        this.cardCustomization.createCustomizationTask
+      ).perform({
+        issuerName: workflowSession.state.issuerName,
+        colorSchemeId: workflowSession.state.colorScheme.id,
+        patternId: workflowSession.state.pattern.id,
       });
+
+      yield taskFor(this.layer2Network.issuePrepaidCard)
+        .perform(this.faceValue, customization.did)
+        .then((address: string) => {
+          this.args.workflowSession.update('prepaidCardAddress', address);
+          this.args.onComplete();
+        });
+    } catch (e) {
+      let insufficientFunds = e.message.startsWith(
+        'Safe not have enough balance to make prepaid card(s).'
+      );
+      let tookTooLong = e.message.startsWith(
+        'Transaction took too long to complete'
+      );
+      if (insufficientFunds) {
+        // We probably want to cancel the workflow at this point
+        // And tell the user to go deposit funds
+        this.error = 'INSUFFICIENT_FUNDS';
+      } else if (tookTooLong) {
+        this.error = 'TIMEOUT';
+      } else {
+        // Basically, for pretty much everything we want to make the user retry or seek support
+        this.error = 'DEFAULT';
+      }
+    }
```

Comment:
@habdelra currently we're catching two errors from the sdk's prepaid card creation and displaying specific error messages for these cases:

When the transaction times out: We tell the user to check with Cardstack support about your transaction (we should give them the transaction hash for this)
When there aren't enough funds: We tell the user to add funds via the deposit workflow
Do you think there are any other cases to catch, or does this cover it? Also, wondering whether you have any thoughts on more structured/specialized errors that we can identify by type/by a shared constant rather than string matching? Having errors like this would make it easy to get the minimum amount required for the prepaid card the user wants, so we can tell the user in the UI.

Reply:
Another kind of error that is potentially possible is creating a prepaid card with unapproved tokens--but that would be impossible to do from the UI, so I think it's ok to let that blow up naturally.

You could also get an error if you attempted to sign the txn from a card wallet that had a different private key than the owner of the safe. I'm not too sure how to replicate this in the real world. Perhaps that is also ok to let blow up naturally.

Yeah error codes would not be a bad idea--where we could throw our own type of error from the SDK that has a `code` property` on it that you could test for instead. I'm totally down for this.